WireAPI/Sources/WireAPI/Network/APIService/APIService.swift

```diff
@@ -99,14 +105,82 @@ public final class APIService: APIServiceProtocol {
         var request = request
 
         if requiringAccessToken {
-            guard let accessToken = authenticationStorage.fetchAccessToken() else {
-                throw APIServiceError.missingAccessToken
-            }
-
+            let accessToken = try await getAccessToken()
             request.setAccessToken(accessToken)
         }
 
         return try await networkService.executeRequest(request)
     }
 
+    private func getAccessToken() async throws -> AccessToken {
+        guard
+            let currentAccessToken = await authenticationStorage.fetchAccessToken(),
+            !currentAccessToken.isExpiring
+        else {
+            let newAccessToken = try await getNewAccessToken()
+            await authenticationStorage.storeAccessToken(newAccessToken)
+            return newAccessToken
+        }
+
+        return currentAccessToken
+    }
+
+    private func getNewAccessToken() async throws -> AccessToken {
+        let cookies = try await authenticationStorage.fetchCookies()
+
+        var request = try URLRequestBuilder(path: "/access")
+            .withQueryItem(name: "client_id", value: clientID)
+            .withMethod(.post)
+            .withAcceptType(.json)
+            .withCookies(cookies)
+            .build()
+
+        if let lastKnownAccessToken = await authenticationStorage.fetchAccessToken() {
+            request.setAccessToken(lastKnownAccessToken)
+        }
+
+        let (data, response) = try await networkService.executeRequest(request)
+
+        return try ResponseParser()
+            .success(code: .ok, type: AccessTokenPayload.self)
+            .failure(code: .forbidden, label: "invalid-credentials", error: APIServiceError.invalidCredentials)
+            .parse(code: response.statusCode, data: data)
+    }
```

Comment:
I think we need to be really careful here to handle race conditions. I _assume_ each time we get a new access token a previous token is invalidated and as far as I know these methods can be called many times in a short period of time. I feel we need some mechanism of queuing or something similar

Reply:
What's a concrete scenario?

- Execute request A with token A, sent
- Execute request B with token A, A is expiring, request new token B, invalidating A.
- Request B reaches server, authentication error response is returned.
- Request B is sent with token B, succeeds

So the possibility is that one or more requests in flight will fail because a new token was created before they reached the server?

Reply:
And the queue mechanism you're referring to would be to prevent reentry?

Reply:
> So the possibility is that one or more requests in flight will fail because a new token was created before they reached the server?

Yes exactly. And my worry is mainly how will the app behave in this situation, especially in a situation where the app makes many requests in a short period of time:

- No/flaky internet
- 5 requests are created
- Internet returns
- Token is expiring
- 5 requests for new tokens are made each succeeding but only the last token remains valid.
- After each token request succeeds it's original request is sent
- Perhaps 4/5 requests fail.

I assume if a request fails for token expiry a new request will be made for a new token. So perhaps we can get into some funny loop between as few a two requests where each keeps requesting a new token and expiring the others token.

I expect that usually things will resolve themselves but I think this situation is hard to reason about. For me it would be much simpler to have some kind of queuing where only one access token refresh is happening at a time. At least this will add simplicity to the users of `getAccessToken()` at the expense of complexity in its implementation.

> And the queue mechanism you're referring to would be to prevent reentry?

Essentially yes. 

Perhaps a simple way to kind of achieve this is to use a specific session for these requests and limit that session to 1 concurrent request.

Reply:
If we are to address this situation it would probably make sense to extract this code into a new object.

Reply:
~Perhaps a simple way to kind of achieve this is to use a specific session for these requests and limit that session to 1 concurrent request.~

No this won't help

Reply:
In a greenfield project what I imagine is that we have a specific object for this probably an `actor` that takes care of proving an non expired access token - it can both fetch from storage or update from network. Internally it would probably makes use of a DispatchQueue as I'm not sure how to achieve this async/await

```swift
actor AccessTokenRepository {

    /// Returns a cached access token if valid. If expired, updates the cached token from the server.
    ///
    /// Blocks if a token is currently being refreshed.
    func accessToken() async throws -> String {}
}
```



```

Reply:
ok with the latest changes I think I've addressed this.

Reply:
Looking good!